### PR TITLE
Add WriteTensorBoard hook test

### LINKED ISSUE #89: Test WriteTensorBoard hook

### PR DESCRIPTION
**Pull request by blazekadam**
closes #89

### CODE REVIEW

File: cxflow_tensorflow/hooks/write_tensorboard.py [hunk 1]
```diff
@@ -99,22 +99,22 @@ def after_epoch(self, epoch_id: int, epoch_data: cx.EpochData) -> None:
...
                 if np.isscalar(value):  # try logging the scalar values
                     result = value
                 elif isinstance(value, dict) and 'mean' in value:  # or the mean
                     result = value['mean']
                 elif isinstance(value, dict) and 'nanmean' in value:  # or the nanmean
                     result = value['nanmean']
-                else:
-                    err_message = 'Variable `{}` in stream `{}` is not scalar and does not contain `mean` or' \
+
+                if type(result) not in [int, float]:
```
Comment (FloopCZ) [inline cxflow_tensorflow/hooks/write_tensorboard.py]: I think we should support other scalar types as well. For instance `np.float32` is a usual scalar type. I suggest to trust  the `np.isscalar` in this case and not check only our selected types (altough `str` can be problematic :-( ).

File: cxflow_tensorflow/hooks/write_tensorboard.py [hunk 2]
```diff
@@ -99,22 +99,22 @@ def after_epoch(self, epoch_id: int, epoch_data: cx.EpochData) -> None:
...
+                    err_message = 'Variable `{}` in stream `{}` is not scalar and does not contain `mean` or ' \
                                   '`nanmean` aggregation'.format(variable, stream_name)
                     if self._on_unknown_type == 'warn':
                         logging.warning(err_message)
-                        result = str(value)
                     elif self._on_unknown_type == 'error':
                         raise ValueError(err_message)
-                    else:
-                        continue
+                    continue
```
Comment (FloopCZ) [inline cxflow_tensorflow/hooks/write_tensorboard.py]: If we only warn, we should probably put the variable in the summary anyway, what was wrong with that?

File: cxflow_tensorflow/hooks/write_tensorboard.py [hunk 3]
```diff
@@ -125,12 +125,13 @@ def after_epoch(self, epoch_id: int, epoch_data: cx.EpochData) -> None:
                         raise KeyError(err_message)
                     elif self._on_missing_variable == 'warn':
                         logging.warning(err_message)
-                    else:
-                        continue
+                    continue
```
Comment (FloopCZ) [inline cxflow_tensorflow/hooks/write_tensorboard.py]: ditto with warning

File: cxflow_tensorflow/tests/hooks/write_tensorboard_test.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
+"""
+Test module for :py:mod:`cxflow_tensorflow.hoooks.write_tensorboard` module.
```
Comment (FloopCZ) [inline cxflow_tensorflow/tests/hooks/write_tensorboard_test.py]: hoooooooks! :-)

File: cxflow_tensorflow/tests/hooks/write_tensorboard_test.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+            _ = WriteTensorBoard(output_dir=self.tmpdir, model=model, flush_secs=42, visualize_graph=True)
+            mocked_writer.assert_called_with(logdir=self.tmpdir, flush_secs=42, graph=model.graph)
+
+            _ = WriteTensorBoard(output_dir=self.tmpdir, model=model)
+            mocked_writer.assert_called_with(logdir=self.tmpdir, flush_secs=10, graph=None)
+
+    def test_write(self):
+        """Test if ``WriteTensorBoard`` writes to its FileWriter."""
+        hook = WriteTensorBoard(output_dir=self.tmpdir, model=self.get_model())
+        with mock.patch.object(tf.summary.FileWriter, 'add_summary') as mocked_ad_summary:
```
Comment (FloopCZ) [inline cxflow_tensorflow/tests/hooks/write_tensorboard_test.py]: maybe `mocked_add_summary` ?

File: cxflow_tensorflow/hooks/write_tensorboard.py (hunk 1 above)
Comment (blazekadam) [reply cxflow_tensorflow/hooks/write_tensorboard.py]: `str` is indeed problematic. With `str` you will get an error (from `tensorflow`) stating that only `int` and `float` types are allowed; hence, this check.

File: cxflow_tensorflow/hooks/write_tensorboard.py (hunk 2 above)
Comment (blazekadam) [reply cxflow_tensorflow/hooks/write_tensorboard.py]: You can add only `int` or `float`.

File: cxflow_tensorflow/hooks/write_tensorboard.py (hunk 3 above)
Comment (blazekadam) [reply cxflow_tensorflow/hooks/write_tensorboard.py]: ditto reply

File: cxflow_tensorflow/tests/hooks/write_tensorboard_test.py (hunk 4 above)
Comment (blazekadam) [reply cxflow_tensorflow/tests/hooks/write_tensorboard_test.py]: hooooooks what?

File: cxflow_tensorflow/tests/hooks/write_tensorboard_test.py (hunk 5 above)
Comment (blazekadam) [reply cxflow_tensorflow/tests/hooks/write_tensorboard_test.py]: indeed

File: cxflow_tensorflow/hooks/write_tensorboard.py (hunk 1 above)
Comment (blazekadam) [reply cxflow_tensorflow/hooks/write_tensorboard.py]: I did not try different types such as tuple or dict; nonetheless, I expect the same error as with `str`.

Comment (FloopCZ) [reply cxflow_tensorflow/hooks/write_tensorboard.py]: I see. In such a case, we should probably update the error message to e.g.
```
'Variable `{}` in stream `{}` has to be of type `int` or `float` (or a `dict` with a key named `mean` or `nanmean` whose corresponding value is of type `int` or `float`).'
```

File: cxflow_tensorflow/tests/hooks/write_tensorboard_test.py (hunk 4 above)
Comment (blazekadam) [reply cxflow_tensorflow/tests/hooks/write_tensorboard_test.py]: ah!, hoooks :D

File: cxflow_tensorflow/hooks/write_tensorboard.py (hunk 2 above)
Comment (FloopCZ) [reply cxflow_tensorflow/hooks/write_tensorboard.py]: ok

File: cxflow_tensorflow/hooks/write_tensorboard.py (hunk 3 above)
Comment (FloopCZ) [reply cxflow_tensorflow/hooks/write_tensorboard.py]: ok

File: cxflow_tensorflow/hooks/write_tensorboard.py (hunk 1 above)
Comment (blazekadam) [reply cxflow_tensorflow/hooks/write_tensorboard.py]: Wow, very verbose! Users will love it!

**Comment by FloopCZ** [comment]
This coverall person is very talkative...

**Comment by blazekadam** [comment]
We can shut his mouth if you will.

**Pull request merged**